extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/TestsAsBeansProcessor.java

```diff
@@ -46,6 +55,36 @@ public void testClassBeans(List<TestClassBeanBuildItem> items, BuildProducer<Add
         producer.produce(builder.build());
     }
 
+    @BuildStep(onlyIf = IsTest.class)
+    AnnotationsTransformerBuildItem vetoTestClassesNotMatchingTestProfile(Optional<TestProfileBuildItem> testProfile) {
+        // Since we only need to register all test classes that belong to the "current" test profile
+        // we have to veto other test classes during the build
+        return new AnnotationsTransformerBuildItem(AnnotationTransformation
+                .forClasses()
+                .when(tc -> {
+                    ClassInfo maybeTestClass = tc.declaration().asClass();
+                    if (maybeTestClass.hasAnnotation(QUARKUS_TEST)
+                            || maybeTestClass.hasAnnotation(QUARKUS_INTEGRATION_TEST)
+                            || maybeTestClass.hasAnnotation(QUARKUS_MAIN_TEST)) {
+                        AnnotationInstance testProfileAnnotation = maybeTestClass.declaredAnnotation(TEST_PROFILE);
+                        if (testProfile.isEmpty()) {
+                            // No test profile set - veto all test classes with a test profile
+                            return testProfileAnnotation != null;
+                        } else {
+                            // Test profile set - veto all test classes with unset/different test profile
+                            if (testProfileAnnotation == null) {
+                                return true;
+                            }
+                            String testProfileClassName = testProfile.get().getTestProfileClassName();
+                            String annotationProfileClassName = testProfileAnnotation.value().asClass().name().toString();
+                            return !testProfileClassName.equals(annotationProfileClassName);
+                        }
+                    }
+                    return false;
+                })
+                .transform(tc -> tc.add(AnnotationInstance.builder(DotNames.VETOED).buildWithTarget(tc.declaration()))));
```

Comment:
It doesn't hurt, but `buildWithTarget(tc.declaration())` isn't be necessary, that's done behind the scenes automatically. Just a note, no need to change.

Reply:
I don't think it does when you build a new `AnnotationInstance` using a builder. You're talking about `org.jboss.jandex.AnnotationOverlayImpl.TransformationContextImpl.add(Class<? extends Annotation>)`, right?

Reply:
Right.